File: frontend/src/types/task.ts
/**
 * 任务监控相关类型定义
 */

export enum TaskStatus {
  PENDING = 'PENDING',
  PROGRESS = 'PROGRESS',
  SUCCESS = 'SUCCESS',
  FAILURE = 'FAILURE',
  RETRY = 'RETRY',
  REVOKED = 'REVOKED'
}

export enum TaskType {
  OCR_PROCESSING = 'ocr_processing',
  LLM_PROCESSING = 'llm_processing',
  KG_EXTRACTION = 'kg_extraction',
  FILE_PROCESSING = 'file_processing',
  EXPORT_PROCESSING = 'export_processing',
  IMPORT_PROCESSING = 'import_processing',
  HEALTH_CHECK = 'health_check'
}

export enum TaskPriority {
  LOW = 'low',
  NORMAL = 'normal',
  HIGH = 'high',
  CRITICAL = 'critical'
}

export interface TaskMonitor {
  id: string;
  task_id: string;
  task_name: string;
  task_type: TaskType;
  status: TaskStatus;
  priority: TaskPriority;
  
  // 进度信息
  progress: number;
  current_step?: string;
  total_steps?: number;
  
  // 执行信息
  user_id: string;
  worker_name?: string;
  queue_name: string;
  
  // 时间信息
  created_at: string;
  started_at?: string;
  completed_at?: string;
  updated_at: string;
  
  // 执行时间统计
  estimated_duration?: number;
  actual_duration?: number;
  execution_time?: number;
  
  // 任务参数和结果
  parameters?: Record<string, any>;
  result?: Record<string, any>;
  error_info?: Record<string, any>;
  
  // 重试信息
  retry_count: number;
  max_retries: number;
  retry_delay: number;
  
  // 状态消息
  message?: string;
  
  // 标记和标签
  tags?: Record<string, any>;
  is_critical: boolean;
  
  // 计算属性
  is_active: boolean;
  is_completed: boolean;
  can_retry: boolean;
}

export interface TaskStatistics {
  total_tasks: number;
  active_tasks: number;
  completed_tasks: number;
  failed_tasks: number;
  success_rate: number;
  average_duration?: number;
  
  // 按类型统计
  by_type: Record<string, number>;
  // 按状态统计
  by_status: Record<string, number>;
  // 按优先级统计
  by_priority: Record<string, number>;
}

export interface TaskQueueInfo {
  queue_name: string;
  active_tasks: number;
  scheduled_tasks: number;
  reserved_tasks: number;
  total_tasks: number;
}

export interface WorkerInfo {
  worker_name: string;
  status: string;
  active_tasks: number;
  processed_tasks: number;
  load_average: number[];
  last_heartbeat?: string;
}

export interface TaskFilterParams {
  status?: TaskStatus[];
  task_type?: TaskType[];
  priority?: TaskPriority[];
  user_id?: string;
  queue_name?: string;
  is_critical?: boolean;
  created_after?: string;
  created_before?: string;
  limit?: number;
  offset?: number;
  order_by?: string;
  order_desc?: boolean;
}

export interface TaskRetryRequest {
  task_id: string;
  reason?: string;
}

export interface TaskBatchOperation {
  task_ids: string[];
  operation: 'cancel' | 'retry' | 'delete';
  reason?: string;
}

export interface TaskWebSocketMessage {
  type: string;
  task_id: string;
  user_id: string;
  status: TaskStatus;
  progress?: number;
  message?: string;
  timestamp: string;
  data?: Record<string, any>;
}

export interface TaskListResponse {
  items: TaskMonitor[];
  total: number;
  limit: number;
  offset: number;
}

// WebSocket 消息类型
export interface WebSocketMessage {
  type: 'connection_established' | 'task_update' | 'subscription_confirmed' | 'unsubscription_confirmed' | 'subscriptions_list' | 'heartbeat_ack' | 'error' | 'pong';
  task_id?: string;
  user_id?: string;
  status?: TaskStatus;
  progress?: number;
  message?: string;
  timestamp: string;
  data?: Record<string, any>;
  task_ids?: string[];
}

// 任务状态颜色映射
export const TASK_STATUS_COLORS: Record<TaskStatus, string> = {
  [TaskStatus.PENDING]: 'blue',
  [TaskStatus.PROGRESS]: 'orange',
  [TaskStatus.SUCCESS]: 'green',
  [TaskStatus.FAILURE]: 'red',
  [TaskStatus.RETRY]: 'purple',
  [TaskStatus.REVOKED]: 'gray'
};

// 任务优先级颜色映射
export const TASK_PRIORITY_COLORS: Record<TaskPriority, string> = {
  [TaskPriority.LOW]: 'gray',
  [TaskPriority.NORMAL]: 'blue',
  [TaskPriority.HIGH]: 'orange',
  [TaskPriority.CRITICAL]: 'red'
};

// 任务类型显示名称
export const TASK_TYPE_NAMES: Record<TaskType, string> = {
  [TaskType.OCR_PROCESSING]: 'OCR处理',
  [TaskType.LLM_PROCESSING]: 'LLM处理',
  [TaskType.KG_EXTRACTION]: '知识图谱抽取',
  [TaskType.FILE_PROCESSING]: '文件处理',
  [TaskType.EXPORT_PROCESSING]: '导出处理',
  [TaskType.IMPORT_PROCESSING]: '导入处理',
  [TaskType.HEALTH_CHECK]: '健康检查'
};

// 任务状态显示名称
export const TASK_STATUS_NAMES: Record<TaskStatus, string> = {
  [TaskStatus.PENDING]: '等待中',
  [TaskStatus.PROGRESS]: '进行中',
  [TaskStatus.SUCCESS]: '成功',
  [TaskStatus.FAILURE]: '失败',
  [TaskStatus.RETRY]: '重试中',
  [TaskStatus.REVOKED]: '已取消'
};

// 任务优先级显示名称
export const TASK_PRIORITY_NAMES: Record<TaskPriority, string> = {
  [TaskPriority.LOW]: '低',
  [TaskPriority.NORMAL]: '普通',
  [TaskPriority.HIGH]: '高',
  [TaskPriority.CRITICAL]: '紧急'
};

File: backend/app/schemas/task.py
"""
任务监控相关的Pydantic模式
"""
from datetime import datetime
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, Field
from uuid import UUID

from app.models.task import TaskStatus, TaskType, TaskPriority


class TaskMonitorBase(BaseModel):
    """任务监控基础模式"""
    task_name: str = Field(..., description="任务名称")
    task_type: TaskType = Field(..., description="任务类型")
    priority: TaskPriority = Field(TaskPriority.NORMAL, description="任务优先级")
    parameters: Optional[Dict[str, Any]] = Field(None, description="任务参数")
    estimated_duration: Optional[int] = Field(None, description="预估执行时间(秒)")
    max_retries: int = Field(3, description="最大重试次数")
    retry_delay: int = Field(60, description="重试延迟(秒)")
    tags: Optional[Dict[str, Any]] = Field(None, description="任务标签")
    is_critical: bool = Field(False, description="是否为关键任务")


class TaskMonitorCreate(TaskMonitorBase):
    """创建任务监控"""
    task_id: str = Field(..., description="Celery任务ID")
    user_id: str = Field(..., description="用户ID")
    queue_name: str = Field("default", description="队列名称")


class TaskMonitorUpdate(BaseModel):
    """更新任务监控"""
    status: Optional[TaskStatus] = None
    progress: Optional[int] = Field(None, ge=0, le=100, description="进度百分比")
    current_step: Optional[str] = None
    total_steps: Optional[int] = None
    worker_name: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    result: Optional[Dict[str, Any]] = None
    error_info: Optional[Dict[str, Any]] = None
    message: Optional[str] = None
    retry_count: Optional[int] = None


class TaskMonitorResponse(TaskMonitorBase):
    """任务监控响应"""
    id: UUID
    task_id: str
    status: TaskStatus
    progress: int
    current_step: Optional[str]
    total_steps: Optional[int]
    user_id: str
    worker_name: Optional[str]
    queue_name: str
    
    created_at: datetime
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    updated_at: datetime
    
    actual_duration: Optional[int]
    result: Optional[Dict[str, Any]]
    error_info: Optional[Dict[str, Any]]
    message: Optional[str]
    
    retry_count: int
    execution_time: Optional[int]
    is_active: bool
    is_completed: bool
    can_retry: bool
    
    class Config:
        from_attributes = True


class TaskStatistics(BaseModel):
    """任务统计信息"""
    total_tasks: int
    active_tasks: int
    completed_tasks: int
    failed_tasks: int
    success_rate: float
    average_duration: Optional[float]
    
    # 按类型统计
    by_type: Dict[str, int]
    # 按状态统计
    by_status: Dict[str, int]
    # 按优先级统计
    by_priority: Dict[str, int]


class TaskQueueInfo(BaseModel):
    """任务队列信息"""
    queue_name: str
    active_tasks: int
    scheduled_tasks: int
    reserved_tasks: int
    total_tasks: int


class WorkerInfo(BaseModel):
    """工作节点信息"""
    worker_name: str
    status: str
    active_tasks: int
    processed_tasks: int
    load_average: List[float]
    last_heartbeat: Optional[datetime]


class TaskRetryRequest(BaseModel):
    """任务重试请求"""
    task_id: str
    reason: Optional[str] = None


class TaskBatchOperation(BaseModel):
    """批量任务操作"""
    task_ids: List[str]
    operation: str  # cancel, retry, delete
    reason: Optional[str] = None


class TaskFilterParams(BaseModel):
    """任务过滤参数"""
    status: Optional[List[TaskStatus]] = None
    task_type: Optional[List[TaskType]] = None
    priority: Optional[List[TaskPriority]] = None
    user_id: Optional[str] = None
    queue_name: Optional[str] = None
    is_critical: Optional[bool] = None
    created_after: Optional[datetime] = None
    created_before: Optional[datetime] = None
    limit: int = Field(50, ge=1, le=1000)
    offset: int = Field(0, ge=0)
    order_by: str = Field("created_at", description="排序字段")
    order_desc: bool = Field(True, description="是否降序")


class TaskWebSocketMessage(BaseModel):
    """WebSocket任务消息"""
    type: str  # task_update, task_created, task_completed, task_failed
    task_id: str
    user_id: str
    status: TaskStatus
    progress: Optional[int] = None
    message: Optional[str] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    data: Optional[Dict[str, Any]] = None